rpc/rpc.go

```diff
@@ -1256,245 +1249,234 @@ func (c *Conn) handleRelease(ctx context.Context, id exportID, count uint32) err
+			return err
 		}
-		// TODO(maybe): check generation?
 
 		// Since this Cap'n Proto RPC implementation does not send imports
 		// unless they are fully dequeued, we can just immediately loop back.
 		id := d.Context().SenderLoopback()
-		err = c.sendMessage(ctx, func(msg rpccp.Message) error {
-			d, err := msg.NewDisembargo()
+		c.sendMessageAsync(ctx, func(m rpccp.Message) error {
+			defer release()
+			defer client.Release()
+
+			d, err := m.NewDisembargo()
 			if err != nil {
-				return err
+				return rpcerr.Annotatef(err, "incoming disembargo: send receiver loopback")
 			}
+
 			tgt, err := d.NewTarget()
 			if err != nil {
-				return err
+				return rpcerr.Annotatef(err, "incoming disembargo: send receiver loopback")
 			}
+
 			tgt.SetImportedCap(uint32(imp.id))
 			d.Context().SetReceiverLoopback(id)
 			return nil
 		})
-		c.mu.Unlock()
-		client.Release()
-		if err != nil {
-			c.report(annotate(err, "incoming disembargo: send receiver loopback"))
-		}
+
 	default:
-		c.report(failedf("incoming disembargo: context %v not implemented", d.Context().Which()))
-		var err error
-		syncutil.With(&c.mu, func() {
-			err = c.sendMessage(ctx, func(msg rpccp.Message) error {
-				mm, err := msg.NewUnimplemented()
-				if err != nil {
-					return err
-				}
-				if err := mm.SetDisembargo(d); err != nil {
-					return err
-				}
-				return nil
-			})
-		})
-		if err != nil {
-			c.report(annotate(err, "incoming disembargo: send unimplemented"))
-		}
-	}
-	return nil
-}
+		c.er.ReportError(fmt.Errorf("incoming disembargo: context %v not implemented", d.Context().Which()))
+		c.sendMessageAsync(ctx, func(m rpccp.Message) (err error) {
+			defer release()
 
-func (c *Conn) handleUnknownMessage(ctx context.Context, recv rpccp.Message) error {
-	c.report(failedf("unknown message type %v from remote", recv.Which()))
-	var err error
-	syncutil.With(&c.mu, func() {
-		err = c.sendMessage(ctx, func(msg rpccp.Message) error {
-			return msg.SetUnimplemented(recv)
+			if m, err = m.NewUnimplemented(); err == nil {
+				err = m.SetDisembargo(d)
+			}
+
+			if err != nil {
+				err = rpcerr.Annotate(err, "incoming disembargo: send unimplemented")
+			}
+
+			return
 		})
-	})
-	if err != nil {
-		c.report(annotate(err, "send unimplemented"))
 	}
+
 	return nil
 }
 
 // startTask increments c.tasks if c is not shutting down.
 // It returns whether c.tasks was incremented.
 //
 // The caller must be holding onto c.mu.
-func (c *Conn) startTask() bool {
-	select {
-	case <-c.bgctx.Done():
-		return false
-	default:
+func (c *Conn) startTask() (ok bool) {
+	if c.bgctx.Err() == nil {
 		c.tasks.Add(1)
-		return true
+		ok = true
 	}
+
+	return
 }
 
-// sendMessage creates a new message on the Sender, calls f to build it,
-// and sends it if f does not return an error.  When f returns, the
-// message must have a nil CapTable.  The caller must be holding onto
-// c.mu.  While f is being called, it will be holding onto the sender
-// lock, but not c.mu.
```

Comment:
It would be good to retain some top level comment for this function, even if it needs adjusting. The roles and invariants around this and the async version are a bit subtle and worth documenting clearly.